README.md
# Rust Reverse Geocoder
A fast reverse geocoder in Rust. Inspired by Python [reverse-geocoder](https://github.com/thampiman/reverse-geocoder).

## Links

- [Crate](https://crates.io/crates/reverse_geocoder)
- [2.0.0 Docs](https://docs.rs/reverse_geocoder/2.0.0/reverse_geocoder/index.html)
- [1.0.1 Docs](https://docs.rs/reverse_geocoder/1.0.1/reverse_geocoder/)
- Build status: [![Build Status](https://travis-ci.org/llambda/rrgeo.svg?branch=master)](https://travis-ci.org/llambda/rrgeo)

## Description 
 
`rrgeo` takes a latitude and longitude as input and returns the closest city, country, latitude, and longitude, using a k-d tree to efficiently find the nearest neighbour based on a known list of locations. This can be useful if you need to reverse geocode a large number of coordinates quickly, or just need the rough location of coordinates but don't want the expense or complication of an online reverse geocoder.

This crate is implemented as a [library](https://crates.io/crates/reverse_geocoder), an [Actix](https://actix.rs/) REST API, an [Iron](https://github.com/iron/iron) REST API, and as a command-line utility, thanks to [Cargo workspaces](https://doc.rust-lang.org/book/ch14-03-cargo-workspaces.html).

## Usage

### Command line search

Example usage:

```
> cargo run -p rrgeo-cmd --release 40 -73
71 ms to load cities.csv
3 ms to build the KdTree
(40.72788, -73.09761): West Sayville New York Suffolk County US
```

### Actix Web Server

Example usage:

```
cargo run -p rrgeo-actix --release
```

Navigate to [the local web server](http://localhost:3000/?lat=40&long=-73).

### Warp Web Server

Example usage:

```
cargo run -p rrgeo-warp --release
```

Navigate to [the local web server](http://localhost:3000/?lat=40&long=-73).

## Benchmarks 

The core library ([measured with criterion](https://github.com/japaric/criterion.rs)):

```
> cargo bench
Benchmarking search: Warming up for 3.0000 s5 ms to build the KdTree
search                  time:   [1.1530 ms 1.1639 ms 1.1748 ms]                    
                        change: [-2.6577% -1.4078% -0.0250%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 100 measurements (1.00%)
  1 (1.00%) high mild

```


Served over [Actix Web](https://actix.rs/):

```
> cargo run --release --bin rrgeo-actix
> wrk --latency http://localhost:3000/\?lat\=40\&long\=\-73
Running 10s test @ http://localhost:3000/?lat=40&long=-73
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.47ms    0.87ms  13.01ms   74.17%
    Req/Sec     2.02k    39.42     2.07k    79.50%
  Latency Distribution
     50%    1.96ms
     75%    3.86ms
     90%    3.93ms
     99%    4.06ms
  40170 requests in 10.00s, 8.58MB read
Requests/sec:   4015.31
Transfer/sec:      0.86MB


```

Served over [Iron](http://ironframework.io/):

```
> cargo run --release --bin rrgeo-iron
> wrk --latency http://localhost:3000/\?lat\=40\&long\=\-73
Running 10s test @ http://localhost:3000/?lat=40&long=-73
  2 threads and 10 connections
  Thread Stats   Avg      Stdev     Max   +/- Stdev
    Latency     2.55ms    1.15ms  25.73ms   76.03%
    Req/Sec     1.99k   110.23     2.20k    82.50%
  Latency Distribution
     50%    1.96ms
     75%    3.35ms
     90%    3.92ms
     99%    5.95ms
  39563 requests in 10.01s, 8.23MB read
Requests/sec:   3954.05
Transfer/sec:    841.78KB

```


## Performance

Below we have comparisons between the Rust, Python and Node.js versions.

|              | Rust | Node   |
|--------------|------|--------|
| Load CSV     | 61ms | 1221ms |
| Build KdTree | 4ms  | 805ms  |
| Search       | 1.1ms  | 0.5ms |

Most of the performance differences appear to be in time taken to load the CSV file and create the k-d tree, but not searching the tree. Searching time resembles algorithmic complexity of [k-d tree](https://en.wikipedia.org/wiki/K-d_tree). Python version is partly implemented in C++ meaning it is not a purely Python implementation. (It might be interesting to see how a pure Python version performs.) The Node.js version is pure JavaScript, as in, not using C add-ons.

### Rust --release build

```
     Running `target/release/web`
PT0.062677465S seconds to load cities.csv
PT0.003835230S seconds to build the KdTree
PT0.068904911S seconds to search
PT0.002596743S seconds to search
PT0.002887542S seconds to search

```

### Rust --debug build

```
     Running `target/debug/web`
PT1.198010357S seconds to load cities.csv
PT0.124435778S seconds to build the KdTree
PT1.401588031S seconds to search
PT0.077837996S seconds to search
PT0.078178297S seconds to search

```

### Python mode 1 (single threaded K-D tree)

```
➜  reverse-geocoder git:(master) ✗ time python mode1.py
Loading formatted geocoded file...
[{'name': 'Saint Louis Park', 'cc': 'US', 'lon': '-93.34801', 'admin1': 'Minnesota', 'admin2': 'Hennepin County', 'lat': '44.9483'}]

python mode1.py  1.60s user 0.22s system 98% cpu 1.847 total
```

### Python mode 2 (multi threaded K-D tree)

```
➜  reverse-geocoder git:(master) ✗ time python mode2.py
Loading formatted geocoded file...
[{'name': 'Saint Louis Park', 'cc': 'US', 'lon': '-93.34801', 'admin1': 'Minnesota', 'admin2': 'Hennepin County', 'lat': '44.9483'}]

python mode2.py  2.82s user 0.34s system 142% cpu 2.221 total
```

### [nreverse](https://github.com/llambda/nreverse) (Node.js version)

```
load modules: 12.619ms
load cities.csv: 1221.833ms
create kdtree: 805.310ms
search tree: 0.758ms
search tree: 0.086ms
search tree: 0.198ms
search tree: 0.104ms
search tree: 0.031ms
total_heap_size 114mb
total_heap_size_executable 5mb
total_physical_size 112mb
total_available_size 1325mb
used_heap_size 83mb
heap_size_limit 1432mb
malloced_memory 0mb
peak_malloced_memory 4mb
does_zap_garbage 0mb

```

## License

Licensed under either of

 * Apache License, Version 2.0, ([LICENSE-APACHE](LICENSE-APACHE) or http://www.apache.org/licenses/LICENSE-2.0)
 * MIT license ([LICENSE-MIT](LICENSE-MIT) or http://opensource.org/licenses/MIT)

at your option.




CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
## [2.0.0] - 2020-5-2
### Changed
- No changes from alpha

## [2.0.0-alpha.0] - 2020-4-4
### Changed
#### Update and improve error handling and function signatures.

* Simplify the `search` API function signature
* Use standard 2018 edition error handling idioms

## [1.0.1] - 2020-3-29
### Added
- Rustdoc documentation
### Changed
- Removed an unnecessary `unwrap` in `Locations::from_path`

## [1.0.0] - 2020-3-28

- First release

src/main.rs
#![feature(test)]
#![feature(phase)]
#![feature(box_syntax)]
extern crate kdtree;
extern crate csv;
extern crate rustc_serialize;
extern crate time;
extern crate params;

use kdtree::KdTree;
use time::PreciseTime;

extern crate iron;
extern crate router;

use iron::prelude::*;
use iron::status;
use rustc_serialize::json;
use router::Router;

#[derive(Clone, RustcDecodable, RustcEncodable)]
struct Record {
    lat: f64,
    lon: f64,
    name: String,
    admin1: String,
    admin2: String,
    admin3: String,
}

pub struct Locations {
    records: Vec<([f64; 2], Record)>,
}

impl Locations {
    fn from_file() -> Locations {
        let start = PreciseTime::now();
        let mut records = Vec::new();

        let mut rdr = csv::Reader::from_file("cities.csv").unwrap();

        for record in rdr.decode() {
            let r: Record = record.unwrap();
            records.push(([r.lat, r.lon], r));
        }

        let end = PreciseTime::now();

        println!("{} seconds to load cities.csv", start.to(end));

        Locations {
            records: records,
        }
    }
}

pub struct ReverseGeocoder<'a> {
    tree: KdTree<'a, &'a Record>,
}

impl<'a> ReverseGeocoder<'a> {
    fn new(loc: &'a Locations) -> ReverseGeocoder<'a> {
        let mut r = ReverseGeocoder::<'a> {
            tree: KdTree::new(2),
        };
        r.initialize(loc);
        r
    }

    fn initialize(&mut self, loc: &'a Locations) {
        let start = PreciseTime::now();
        for record in &loc.records {
            self.tree.add(&record.0, &record.1).unwrap();
        }
        let end = PreciseTime::now();
        println!("{} seconds to build the KdTree", start.to(end));
    }

    fn search(&self, loc: &[f64; 2]) -> Option<Record> {
        use kdtree::distance::squared_euclidean;

        let y = self.tree.nearest(loc, 1, &squared_euclidean).unwrap();

        if y.len() > 0 {
            return Some((*y[0].1).clone());
        } else {
            return None;
        }
    }

}

fn print_record(r: &Record) {
    println!("({}, {}): {} / {} / {} / {}", r.lat, r.lon, r.name, r.admin1, r.admin2, r.admin3);
}

fn main() {
    let loc = Locations::from_file();
    let geocoder = ReverseGeocoder::new(&loc);

    let test = geocoder.search(&[44.962786, -93.344722]).unwrap();

    print_record(&test);
    let mut router = Router::new();

    router.get("/reverse", hello_world);

    fn hello_world(_: &mut Request) -> IronResult<Response> {
        use params::{Map, Params};

        let map: Map = try!(req.get_ref::<Params>());

        let loc = Locations::from_file();
        let geocoder = ReverseGeocoder::new(&loc);
        let y = geocoder.search(&[44.962786, -93.344722]).unwrap();
        let payload = json::encode(&y).unwrap();
        Ok(Response::with((status::Ok, payload)))
    }

    Iron::new(router).http("localhost:3000").unwrap();
    println!("On 3000");
}

extern crate test;

mod tests {

    #[test]
    fn it_works() {
        let loc = super::Locations::from_file();
        let geocoder = super::ReverseGeocoder::new(&loc);
        let y = geocoder.search(&[44.962786, -93.344722]);
        assert_eq!(y.is_some(), true);
        let slp = y.unwrap();

        assert_eq!(slp.name, "Saint Louis Park");

        // [44.894519, -93.308702] is 60 St W @ Penn Ave S, Minneapolis, Minnesota; however, this is physically closer to Richfield
        let mpls = geocoder.search(&[44.894519, -93.308702]).unwrap();
        assert_eq!(mpls.name, "Richfield");

        // [44.887055, -93.334204] is HWY 62 and Valley View Road, whish is in Edina
        let edina = geocoder.search(&[44.887055, -93.334204]).unwrap();
        assert_eq!(edina.name, "Edina");
    }

    // #[bench]
    // fn bench_lookup(b: &mut Bencher) {
    //
    // }
}


/*
fn geodetic_in_ecef(geo_coords: (f32, f32)) -> (f32, f32, f32) {
    let a = 6378.137; // major axis in kms
    let e2 = 0.00669437999014;

    let lat = geo_coords.0;
    let lon = geo_coords.1;

    let lat_r = lat.to_radians();
    let lon_r = lon.to_radians();
    let normal = a / (1f32 - e2 * lat_r.sin().powi(2));

    let x = normal * lat_r.cos() * lon_r.cos();
    let y = normal * lat_r.cos() * lon_r.sin();
    let z = normal * (1f32 - e2) * lat.sin();

    (x, y, z)
}
*/


src/geocoder.rs

use kdtree::KdTree;
use kdtree::distance::squared_euclidean;
use time::PreciseTime;
extern crate quick_csv;

extern crate rocket;
extern crate serde_json;
extern crate rustc_serialize;

#[derive(Debug, Clone, Serialize, Deserialize, RustcDecodable, RustcEncodable)]
pub struct Record {
    lat: f64,
    lon: f64,
    name: String,
    admin1: String,
    admin2: String,
    admin3: String,
}

pub struct Locations {
    records: Vec<([f64; 2], Record)>,
}

impl Locations {
    pub fn from_file() -> Locations {
        let start = PreciseTime::now();
        let mut records = Vec::new();

        let reader = quick_csv::Csv::from_file("cities.csv").unwrap().has_header(true);

        for read_record in reader {
            let record: Record = read_record.unwrap().decode().unwrap();
            records.push(([record.lat, record.lon], record));
        }

        let end = PreciseTime::now();

        println!("{} ms to load cities.csv", start.to(end).num_milliseconds());

        Locations { records: records }
    }
}

pub struct ReverseGeocoder<'a> {
    tree: KdTree<&'a Record, &'a [f64; 2]>,
}

impl<'a> ReverseGeocoder<'a> {
    pub fn new(loc: &'a Locations) -> ReverseGeocoder<'a> {
        let mut reverse_geocoder =
            ReverseGeocoder::<'a> { tree: KdTree::new_with_capacity(2, loc.records.len()) };
        reverse_geocoder.initialize(loc);
        reverse_geocoder
    }

    fn initialize(&mut self, loc: &'a Locations) {
        let start = PreciseTime::now();
        for record in &loc.records {
            self.tree.add(&record.0, &record.1).unwrap();
        }
        let end = PreciseTime::now();
        println!("{} ms to build the KdTree", start.to(end).num_milliseconds());
    }

    pub fn search(&'a self, loc: &[f64; 2]) -> Option<&'a Record> {
        let nearest = self.tree.nearest(loc, 1, &squared_euclidean).unwrap();
        if nearest.is_empty() {
            None
        } else {
            Some(&nearest[0].1)
        }
    }
}

pub fn print_record(record: &Record) {
    println!("({}, {}): {} {} {} {}",
             record.lat,
             record.lon,
             record.name,
             record.admin1,
             record.admin2,
             record.admin3);
}

mod tests {
    use super::*;

    #[test]
    fn it_works() {
        let loc = Locations::from_file();
        let geocoder = ReverseGeocoder::new(&loc);
        let y = geocoder.search(&[44.962786, -93.344722]);
        assert_eq!(y.is_some(), true);
        let slp = y.unwrap();

        assert_eq!(slp.name, "Saint Louis Park");

        // [44.894519, -93.308702] is 60 St W @ Penn Ave S, Minneapolis, Minnesota; however, this is physically closer to Richfield
        let mpls = geocoder.search(&[44.894519, -93.308702]).unwrap();
        assert_eq!(mpls.name, "Richfield");

        // [44.887055, -93.334204] is HWY 62 and Valley View Road, whish is in Edina
        let edina = geocoder.search(&[44.887055, -93.334204]).unwrap();
        assert_eq!(edina.name, "Edina");
    }
}

src/webserver.rs
#![cfg_attr(feature="clippy", plugin(clippy))]
#![feature(plugin, proc_macro, custom_derive)]
#![plugin(rocket_codegen)]

#[macro_use] extern crate lazy_static;

extern crate rocket;
extern crate serde_json;
#[macro_use] extern crate serde_derive;

extern crate rustc_serialize;

extern crate kdtree;
extern crate time;

use time::PreciseTime;
use rocket::response::content::JSON;

mod geocoder;
use geocoder::Locations;
use geocoder::ReverseGeocoder;

#[derive(FromForm)]
struct LatLongQuery {
    lat: f64,
    lng: f64
}

lazy_static! {
    static ref LOCATIONS: Locations = Locations::from_file();
    static ref GEOCODER: ReverseGeocoder<'static> = ReverseGeocoder::new(&LOCATIONS);
}

#[get("/api/lookup?<latLong>")]
fn lookup(latLong: LatLongQuery) -> JSON<String> {
    let start = PreciseTime::now();
    let y = GEOCODER.search(&[latLong.lat, latLong.lng]).unwrap();
    let end = PreciseTime::now();
    println!("{} ms to search", start.to(end).num_milliseconds());

    // Ok(Response::with((status::Ok, json::encode(y).unwrap())))
    JSON(serde_json::to_string(&y).unwrap())
}


// fn geocoder_middleware(request: &mut Request) -> IronResult<Response> {
//     match request.url.query().clone() {
//         Some(query) => {
//             // println!("{:?}", query);
//             let data = parse(&query).unwrap();
//             // println!("{:?}", data);
//             // println!("{:?}", data.is_object());

//             let obj = data.as_object().unwrap();
//             let lat = obj.get("lat").unwrap().as_str().unwrap().parse::<f64>().unwrap();
//             let long = obj.get("long").unwrap().as_str().unwrap().parse::<f64>().unwrap();

//             let start = PreciseTime::now();
//             let y = GEOCODER.search(&[lat, long]).unwrap();
//             let end = PreciseTime::now();
//             println!("{} ms to search", start.to(end).num_milliseconds());

//             Ok(Response::with((status::Ok, json::encode(y).unwrap())))
//         }
//         None => Ok(Response::with((status::BadRequest, "Need a lat/long"))),
//     }
// }

fn main() {
    // Iron::new(geocoder_middleware).http("localhost:3000").unwrap();
    rocket::ignite().mount("/", routes![lookup]).launch()
    // println!("On 3000");
}


cmd/src/cmdline.rs
use reverse_geocoder::{Locations, ReverseGeocoder, SearchResult};
use std::{env, process::exit};
use std::error::Error;
use time::Instant;

fn main() -> Result<(), Box<dyn Error>> {
    let args: Vec<_> = env::args().collect();

    if args.len() < 3 {
        eprint!("Usage: rreverse lat long\n e.g., rreverse 40.7831 -73.9712\n\n");
        exit(1);
    }

    let lat = args[1].parse::<f64>()?;
    let long = args[2].parse::<f64>()?;

    let loc = Locations::from_memory();
    let geocoder = ReverseGeocoder::new(&loc);

    let start = Instant::now();
    let search_result: SearchResult = geocoder.search((lat, long)).expect("Nothing found.");
    eprintln!("{} ms to search", start.elapsed().whole_milliseconds());

    println!("Location: {}", search_result.record);
    println!("Distance: {}", search_result.distance);

    Ok(())
}


reverse-geocoder/src/reverse_geocoder.rs
#![deny(missing_docs)]
//! A library for fast, offline reverse geocoding. The location data are from [GeoNames](http://www.geonames.org/).
//!
//! # Usage
//! ```
//! use reverse_geocoder::{Locations, ReverseGeocoder, SearchResult};
//!
//! fn main() {
//!     let loc = Locations::from_memory();
//!     let geocoder = ReverseGeocoder::new(&loc);
//!     let coords = (40.7831, -73.9712);
//!     let search_result = geocoder.search(coords).unwrap();
//!     println!("Distance {}", search_result.distance);
//!     println!("Record {}", search_result.record);
//! }
//!```

use kdtree::{distance::squared_euclidean, KdTree};
// use time::Instant;
use serde_derive::{Serialize, Deserialize};
use std::error;
use std::fmt;
use std::path::Path;

/// A parsed location.
#[derive(Debug, Clone, RustcDecodable, RustcEncodable, Serialize, Deserialize)]
pub struct Record {
    /// Latitude
    pub lat: f64,
    /// Longitude
    pub lon: f64,
    /// Location name
    pub name: String,
    /// Administrative district 1
    pub admin1: String,
    /// Administrative district 2
    pub admin2: String,
    /// Administrative district 3
    pub admin3: String,
}

/// Search result from querying a lat/long.
#[derive(Debug, Serialize, Clone)]
pub struct SearchResult<'a> {
    /// Distance away from given lat/long.
    pub distance: f64,
    /// Closest place information.
    pub record: &'a Record,
}

/// A set of location records.
pub struct Locations {
    records: Vec<([f64; 2], Record)>,
}

impl Locations {
    /// Use the built-in cities.csv.
    pub fn from_memory() -> Locations {
        let mut records = Vec::new();
        let my_str = include_str!("../cities.csv");
        let reader = quick_csv::Csv::from_string(my_str).has_header(true);

        for read_record in reader {
            let record: Record = read_record.unwrap().decode().unwrap();
            records.push(([record.lat, record.lon], record));
        }
        Locations { records: records }
    }

    /// Supply your own path to a CSV file.
    pub fn from_path<P: AsRef<Path>>(file_path: P) -> Result<Locations, Box<dyn error::Error>> {
        // let start_load = Instant::now();
        let mut records = Vec::new();

        let reader = quick_csv::Csv::from_file(file_path)?.has_header(true);

        for read_record in reader {
            let record: Record = read_record?.decode()?;
            records.push(([record.lat, record.lon], record));
        }

        // eprintln!("{} ms to load csv", start_load.elapsed().whole_milliseconds());
        Ok(Locations { records: records })
    }
}

/// A reverse geocoder.
pub struct ReverseGeocoder<'a> {
    tree: KdTree<f64, &'a Record, &'a [f64; 2]>,
}

impl<'a> ReverseGeocoder<'a> {
    /// Create a new reverse geocoder from a set of locations.
    pub fn new(loc: &'a Locations) -> ReverseGeocoder<'a> {
        let mut reverse_geocoder = ReverseGeocoder::<'a> {
            tree: KdTree::with_capacity(2, loc.records.len()),
        };
        reverse_geocoder.initialize(loc);
        reverse_geocoder
    }

    fn initialize(&mut self, loc: &'a Locations) {
        // let start = Instant::now();
        for record in &loc.records {
            self.tree.add(&record.0, &record.1).unwrap();
        }
        // let end = Instant::now();
        // println!("{} ms to build the KdTree", (end - start).whole_milliseconds());
    }

    /// Search for the closest record to a given (latitude, longitude). Non-finite numbers will always return None.
    pub fn search(&self, loc: (f64, f64)) -> Option<SearchResult> {
        let nearest = match self.tree.nearest(&[loc.0, loc.1], 1, &squared_euclidean) {
            Ok(nearest) => nearest,
            Err(error) => match error {
                kdtree::ErrorKind::WrongDimension => {
                    panic!("Internal error, kdtree::ErrorKind::WrongDimension should never occur")
                }
                kdtree::ErrorKind::NonFiniteCoordinate => return None,
                kdtree::ErrorKind::ZeroCapacity => {
                    panic!("Internal error, kdtree::ErrorKind::ZeroCapacity should never occur")
                }
            },
        };
        match nearest.get(0) {
            Some(nearest) => Some(SearchResult {
                distance: nearest.0,
                record: nearest.1,
            }),
            None => None,
        }
    }
}

impl fmt::Display for Record {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "({}, {}): {}, {}, {}, {}",
            self.lat, self.lon, self.name, self.admin1, self.admin2, self.admin3
        )
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn it_finds_4_places() {
        let loc = Locations::from_memory();
        let geocoder = ReverseGeocoder::new(&loc);

        let slp = geocoder.search((40.7831, -73.9712)).unwrap();
        assert_eq!(slp.record.name, "Manhattan");

        let slp = geocoder.search((44.962786, -93.344722)).unwrap();

        assert_eq!(slp.record.name, "Saint Louis Park");

        // [44.894519, -93.308702] is 60 St W @ Penn Ave S, Minneapolis, Minnesota; however, this is physically closer to Richfield
        let mpls = geocoder.search((44.894519, -93.308702)).unwrap();
        assert_eq!(mpls.record.name, "Richfield");

        // [44.887055, -93.334204] is HWY 62 and Valley View Road, whish is in Edina
        let edina = geocoder.search((44.887055, -93.334204)).unwrap();
        assert_eq!(edina.record.name, "Edina");
    }

    #[test]
    fn it_loads_locations_from_a_path() -> Result<(), Box<dyn error::Error>> {
        let loc = Locations::from_path("./cities.csv")?;
        let geocoder = ReverseGeocoder::new(&loc);
        let search_result = geocoder.search((45.0, 54.0));
        assert!(search_result.is_some());
        Ok(())
    }

    #[test]
    fn it_loads_locations_from_a_nearly_blank_file() -> Result<(), Box<dyn error::Error>> {
        let loc = Locations::from_path("./nearly-blank.csv")?;
        let geocoder = ReverseGeocoder::new(&loc);
        let search_result = geocoder.search((45.0, 54.0));
        assert!(search_result.is_none());
        Ok(())
    }

    #[test]
    fn it_loads_locations_from_a_blank_file() -> Result<(), Box<dyn error::Error>> {
        let loc = Locations::from_path("./blank.csv")?;
        let geocoder = ReverseGeocoder::new(&loc);
        let search_result = geocoder.search((45.0, 54.0));
        assert!(search_result.is_none());
        Ok(())
    }

    #[test]
    fn it_returns_none_given_an_infinite_coordinate() {
        let loc = Locations::from_memory();
        let geocoder = ReverseGeocoder::new(&loc);
        let search_result = geocoder.search((std::f64::INFINITY, 54.0));
        assert!(search_result.is_none());
    }
}


actix/src/actix.rs
use actix_web::{http, middleware, web, App, HttpResponse, HttpServer, Result};
use lazy_static::lazy_static;
use reverse_geocoder::{Locations, Record, ReverseGeocoder};
use serde_derive::Deserialize;
use std::fmt;

#[derive(Debug)]
enum ReverseGeocodeWebError {
    NotFound,
}

impl fmt::Display for ReverseGeocodeWebError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            ReverseGeocodeWebError::NotFound => write!(f, "Not found"),
        }
    }
}

impl actix_web::error::ResponseError for ReverseGeocodeWebError {
    fn error_response(&self) -> HttpResponse {
        match self {
            ReverseGeocodeWebError::NotFound => HttpResponse::new(http::StatusCode::NOT_FOUND),
        }
    }
}

#[derive(Deserialize)]
struct LatLong {
    lat: f64,
    long: f64,
}

async fn index(lat_long: web::Query<LatLong>) -> Result<web::Json<Record>, ReverseGeocodeWebError> {
    lazy_static! {
        static ref LOCATIONS: Locations = Locations::from_memory();
        static ref GEOCODER: ReverseGeocoder<'static> = ReverseGeocoder::new(&LOCATIONS);
    }

    let search_result = match GEOCODER.search((lat_long.lat, lat_long.long)) {
        Some(result) => result,
        None => return Err(ReverseGeocodeWebError::NotFound),
    };

    Ok(web::Json(search_result.record.clone()))
}

#[actix_rt::main]
async fn main() -> std::io::Result<()> {
    std::env::set_var("RUST_LOG", "actix_web=info");
    env_logger::init();

    HttpServer::new(|| {
        App::new()
            .wrap(middleware::Logger::default())
            .route("/", web::get().to(index))
    })
    .keep_alive(10)
    .bind("127.0.0.1:3000")?
    .run()
    .await
}

#[cfg(test)]
mod tests {
    use super::*;
    extern crate bytes;

    use actix_web::dev::Service;
    use actix_web::{http, test, web, App};

    #[actix_rt::test]
    async fn it_serves_results_on_actix() -> Result<(), ReverseGeocodeWebError> {
        let mut app = test::init_service(App::new().route("/", web::get().to(index))).await;

        let req = test::TestRequest::get()
            .uri("/?lat=44.962786&long=-93.344722")
            .to_request();

        let resp = app.call(req).await.unwrap();

        assert_eq!(resp.status(), http::StatusCode::OK);

        let response_body = match resp.response().body().as_ref() {
            Some(actix_web::body::Body::Bytes(bytes)) => bytes,
            _ => panic!("Response error"),
        };

        assert_eq!(
            response_body,
            r##"{"lat":44.9483,"lon":-93.34801,"name":"Saint Louis Park","admin1":"Minnesota","admin2":"Hennepin County","admin3":"US"}"##
        );

        Ok(())
    }
}
